### PR TITLE
Buy and Sell Transaction Parser

### LINKED ISSUE #130: Implement Transaction Command Parser

### LINKED ISSUE #131: Implement Transaction Command Parser Test Case

### PR DESCRIPTION
**Pull request by Jo3LW**

### CODE REVIEW

File: docs/UserGuide.md [hunk 1]
```diff
@@ -262,6 +262,24 @@ _Details coming soon ..._
 
 ## Command summary
 
+<<<<<<< HEAD
```
Comment (notle1706) [inline docs/UserGuide.md]: I think this is from a merge conflict. Please resolve it

File: docs/UserGuide.md [hunk 2]
```diff
@@ -278,4 +296,5 @@ _Details coming soon ..._
 | **Sort**        | Coming Soon                                                                                                                                                                                                                                                                |
 | **Transaction** | `transaction coy/COMPANY_NAME g/GOODS q/QUANTITY pr/PRICE d/BUY` <br/> e.g `transaction coy/McDonalds g/apples q/100 pr/1.5 d/buy` <br/> `transaction coy/COMPANY_NAME g/GOODS q/QUANTITY pr/PRICE d/SELL` <br/> e.g. `transaction coy/KFC g/Chickens q/50 pr/2.55 d/SELL` |
 | **User Guide**  | `user_guide`                                                                                                                                                                                                                                                               |
+>>>>>>> 78b6ac43588d9cbbfef5505a01a268dd1c451564
```
Comment (notle1706) [inline docs/UserGuide.md]: This is part of the merge conflict as well. Please resolve.

File: src/main/java/seedu/address/logic/commands/SellCommand.java [hunk 3]
```diff
@@ -54,6 +54,8 @@ public SellCommand(Index index, Transaction transaction) {
         this.index = index;
         this.transaction = transaction;
 
+
+
```
Comment (notle1706) [inline src/main/java/seedu/address/logic/commands/SellCommand.java:58]: Perhaps it would be better to remove the empty lines.

File: docs/UserGuide.md [hunk 4]
```diff
@@ -262,6 +262,24 @@ _Details coming soon ..._
...
+| **Find**       | `find KEYWORD [MORE_KEYWORDS]`<br> e.g., `find MacDonalds`                                                                 |
+| **Filter**     | Coming Soon                                                                                                                |
+| **List**       | `list`                                                                                                                     |
+| **Help**       | `help` or `help [COMMAND]` <br> e.g.,`help` or `help add` or `help sort`                                                   |
+| **Sort**       | Coming Soon                                                                                                                |
+| **Buy**        | `buy INDEX g/GOODS q/QUANTITY price/PRICE` <br/> e.g `buy 2 g/apples q/100 price/1.5`                                      |
+| **Sell**       | `sell INDEX g/GOODS q/QUANTITY price/PRICE` <br/> e.g `sell 2 g/apples q/100 price/1.5`                                    |
+| **User Guide** | `user_guide`                                                                                                               |
+=======
```
Comment (notle1706) [inline docs/UserGuide.md:282]: Please remove this line.

File: docs/UserGuide.md (hunk 2 above)
Comment (notle1706) [inline docs/UserGuide.md]: Please remove this line after resolving the merge conflicts.

File: docs/UserGuide.md (hunk 4 above)
Comment (notle1706) [inline docs/UserGuide.md]: There might be conflicts yet to be resolved. Please resolve them.

File: src/main/java/seedu/address/logic/commands/SellCommand.java (hunk 3 above)
Comment (notle1706) [inline src/main/java/seedu/address/logic/commands/SellCommand.java]: Perhaps these 2 lines could be removed.

File: src/main/java/seedu/address/logic/commands/SellCommand.java [hunk 5]
```diff
@@ -71,6 +73,7 @@ public CommandResult execute(Model model) throws CommandException {
         Company editedCompany = companyToEdit;
         editedCompany.addTransaction(transaction);
         model.setCompany(companyToEdit, editedCompany);
+        model.updateFilteredCompanyList(PREDICATE_SHOW_ALL_COMPANIES);
```
Comment (notle1706) [inline src/main/java/seedu/address/logic/commands/SellCommand.java]: Perhaps this line should be deleted in accordance with #151.

File: src/main/java/seedu/address/logic/parser/BuyCommandParser.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+        Goods goods = ParserUtil.parseGoods(argMultimap.getValue(PREFIX_GOODS).orElse(""));
+        Price price = ParserUtil.parsePrice(argMultimap.getValue(PREFIX_PRICE).orElse(""));
+        Quantity quantity = ParserUtil.parseQuantity(argMultimap.getValue(PREFIX_QUANTITY).orElse(""));
+
+        Transaction transaction = new BuyTransaction(goods, price, quantity);
+
+        return new BuyCommand(index, transaction);
+    }
+}
```
Comment (notle1706) [inline src/main/java/seedu/address/logic/parser/BuyCommandParser.java:54]: This class is well-written. Good job!

File: src/main/java/seedu/address/logic/parser/SellCommandParser.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        if (argMultimap.getValue(PREFIX_GOODS).isEmpty() || argMultimap.getValue(PREFIX_PRICE).isEmpty()
+                || argMultimap.getValue(PREFIX_QUANTITY).isEmpty()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, SellCommand.MESSAGE_USAGE));
+        }
+
+        Goods goods = ParserUtil.parseGoods(argMultimap.getValue(PREFIX_GOODS).orElse(""));
+        Price price = ParserUtil.parsePrice(argMultimap.getValue(PREFIX_PRICE).orElse(""));
+        Quantity quantity = ParserUtil.parseQuantity(argMultimap.getValue(PREFIX_QUANTITY).orElse(""));
+
+
```
Comment (notle1706) [inline src/main/java/seedu/address/logic/parser/SellCommandParser.java:48]: Perhaps this line could be deleted.

File: src/main/java/seedu/address/model/company/Company.java [hunk 8]
```diff
@@ -61,6 +61,8 @@ public Name getName() {
         return name;
     }
 
+
+
```
Comment (notle1706) [inline src/main/java/seedu/address/model/company/Company.java]: Perhaps these two lines could be deleted.

File: src/main/java/seedu/address/model/company/Company.java [hunk 9]
```diff
@@ -175,12 +177,9 @@ public String toString() {
         }
 
         TransactionLog transactions = getTransactions();
-        builder.append("; Total transactions: $");
         if (!isNull(transactions) && !transactions.isEmpty()) {
             System.out.println(transactions.calculateNetTransacted());
-            builder.append(transactions.calculateNetTransacted());
-        } else {
-            builder.append("0");
+            builder.append("; Total transactions: $").append(transactions.calculateNetTransacted());
```
Comment (notle1706) [inline src/main/java/seedu/address/model/company/Company.java:182]: Please revert this change as this could induce a bug.

File: src/main/java/seedu/address/model/transaction/TransactionLog.java [hunk 10]
```diff
@@ -38,6 +38,9 @@ public List<Transaction> getTransactionLog() {
      */
     public double calculateNetTransacted() {
         double result = 0;
+        if (transactionList.isEmpty()) {
+            return result;
+        }
```
Comment (notle1706) [inline src/main/java/seedu/address/model/transaction/TransactionLog.java]: Perhaps these lines could be deleted since result would be 0 if the transactionList is empty.

File: src/test/java/seedu/address/model/transaction/BuyTransactionTest.java [hunk 11]
```diff
@@ -31,6 +31,7 @@ public void testToString() {
         Quantity quantity = new Quantity("10");
 
         Transaction transaction = new BuyTransaction(goods, price, quantity);
+        System.out.println(transaction.toString());
```
Comment (notle1706) [inline src/test/java/seedu/address/model/transaction/BuyTransactionTest.java]: Perhaps this line could be deleted.

**Review by notle1706 — CHANGES_REQUESTED** [review]
There are a few codes to change. Aside from that, the Transaction Parser looks good!

File: src/main/java/seedu/address/logic/commands/SellCommand.java [hunk 12]
```diff
@@ -4,6 +4,7 @@
 import static seedu.address.logic.parser.CliSyntax.PREFIX_GOODS;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_PRICE;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_QUANTITY;
+import static seedu.address.model.Model.PREDICATE_SHOW_ALL_COMPANIES;
```
Comment (notle1706) [inline src/main/java/seedu/address/logic/commands/SellCommand.java]: This import would have to be removed in accordance with #151.

File: src/test/java/seedu/address/logic/parser/SellCommandParserTest.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        // only quantity
+        assertParseFailure(parser, "1" + VALID_QUANTITY_SELL_PAPAYA,
+                MESSAGE_INVALID_FORMAT);
+        // only goods
+        assertParseFailure(parser, "1" + VALID_GOODS_SELL_PAPAYA, MESSAGE_INVALID_FORMAT);
+        // goods and price but no quantity
+        assertParseFailure(parser, "1" + VALID_GOODS_SELL_PAPAYA + VALID_PRICE_SELL_PAPAYA,
+               MESSAGE_INVALID_FORMAT);
+
+
```
Comment (ElginL) [inline src/test/java/seedu/address/logic/parser/SellCommandParserTest.java]: Please remove these empty lines

**Review by ElginL — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/model/company/Company.java [hunk 14]
```diff
@@ -175,12 +177,9 @@ public String toString() {
         }
 
         TransactionLog transactions = getTransactions();
-        builder.append("; Total transactions: $");
         if (!isNull(transactions) && !transactions.isEmpty()) {
             System.out.println(transactions.calculateNetTransacted());
```
Comment (notle1706) [inline src/main/java/seedu/address/model/company/Company.java]: Perhaps this line should be removed.

**Review by notle1706 — CHANGES_REQUESTED** [review]
Some changes to be made.

File: src/main/java/seedu/address/model/company/Company.java (hunk 14 above)
Comment (Jo3LW) [reply src/main/java/seedu/address/model/company/Company.java]: Okay will do thank you :)

File: src/main/java/seedu/address/logic/commands/SellCommand.java (hunk 12 above)
Comment (Jo3LW) [reply src/main/java/seedu/address/logic/commands/SellCommand.java]: Okay will do thank you :)

File: src/test/java/seedu/address/logic/parser/SellCommandParserTest.java (hunk 13 above)
Comment (Jo3LW) [reply src/test/java/seedu/address/logic/parser/SellCommandParserTest.java]: Okay will do thank you :)

File: src/test/java/seedu/address/model/transaction/BuyTransactionTest.java (hunk 11 above)
Comment (Jo3LW) [reply src/test/java/seedu/address/model/transaction/BuyTransactionTest.java]: Okay will do thank you :)

File: src/main/java/seedu/address/model/transaction/TransactionLog.java (hunk 10 above)
Comment (Jo3LW) [reply src/main/java/seedu/address/model/transaction/TransactionLog.java]: Okay will do thank you :)

File: src/main/java/seedu/address/logic/parser/BuyCommandParser.java (hunk 6 above)
Comment (Jo3LW) [reply src/main/java/seedu/address/logic/parser/BuyCommandParser.java:54]: Thank you :)

File: src/main/java/seedu/address/logic/commands/SellCommand.java (hunk 5 above)
Comment (Jo3LW) [reply src/main/java/seedu/address/logic/commands/SellCommand.java]: Okay will do thank you :)

File: docs/UserGuide.md (hunk 4 above)
Comment (Jo3LW) [reply docs/UserGuide.md]: Okay will do thank you :)

File: docs/UserGuide.md (hunk 2 above)
Comment (Jo3LW) [reply docs/UserGuide.md]: Okay will do thank you :)

Comment (Jo3LW) [reply docs/UserGuide.md]: Okay will do thank you :)

File: docs/UserGuide.md (hunk 1 above)
Comment (Jo3LW) [reply docs/UserGuide.md]: Okay will do thank you :)

File: src/main/java/seedu/address/model/transaction/BuyTransaction.java [hunk 15]
```diff
@@ -25,4 +25,24 @@ public String toString() {
...
+     */
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        if (!(other instanceof BuyTransaction)) {
+            return false;
+        }
```
Comment (ElginL) [inline src/main/java/seedu/address/model/transaction/BuyTransaction.java:41]: I think this part is testable, but isn't covered by a test yet (as warned by codecov). Perhaps, it might be better if we could add such a case in?

File: src/main/java/seedu/address/model/transaction/SellTransaction.java [hunk 16]
```diff
@@ -24,4 +24,24 @@ public String toString() {
...
+     */
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        if (!(other instanceof SellTransaction)) {
+            return false;
+        }
```
Comment (ElginL) [inline src/main/java/seedu/address/model/transaction/SellTransaction.java:40]: Same as above (testing)

File: src/test/java/seedu/address/logic/parser/SellCommandParserTest.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+        // invalid arguments being parsed as preamble
+        assertParseFailure(parser, "1 some random string", MESSAGE_INVALID_FORMAT);
+
+        // invalid prefix being parsed as preamble
+        assertParseFailure(parser, "1 i/ string", MESSAGE_INVALID_FORMAT);
+    }
+
+
+    //DONE
```
Comment (ElginL) [inline src/test/java/seedu/address/logic/parser/SellCommandParserTest.java]: Not sure what this comment is for 🤔

File: src/test/java/seedu/address/logic/parser/BuyCommandParserTest.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+        // invalid arguments being parsed as preamble
+        assertParseFailure(parser, "1 some random string", MESSAGE_INVALID_FORMAT);
+
+        // invalid prefix being parsed as preamble
+        assertParseFailure(parser, "1 i/ string", MESSAGE_INVALID_FORMAT);
+    }
+
+
+    //DONE
```
Comment (ElginL) [inline src/test/java/seedu/address/logic/parser/BuyCommandParserTest.java]: Would this comment be useful? Or do you think it can be removed 🤔

**Review by ElginL — CHANGES_REQUESTED** [review]
Looking good, just a few minor changes would make it perfect 💯. Good job.

File: src/test/java/seedu/address/logic/parser/SellCommandParserTest.java (hunk 17 above)
Comment (Jo3LW) [reply src/test/java/seedu/address/logic/parser/SellCommandParserTest.java]: Thanks for spotting this ! Will make the necessary changes.

File: src/test/java/seedu/address/logic/parser/BuyCommandParserTest.java (hunk 18 above)
Comment (Jo3LW) [reply src/test/java/seedu/address/logic/parser/BuyCommandParserTest.java]: Thanks for spotting this ! Will make the necessary changes.

**Review by ElginL — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/logic/parser/JeeqTrackerParser.java [hunk 19]
```diff
@@ -16,10 +17,12 @@
 import seedu.address.logic.commands.FindCommand;
 import seedu.address.logic.commands.HelpCommand;
 import seedu.address.logic.commands.ListCommand;
+import seedu.address.logic.commands.SellCommand;
 import seedu.address.logic.commands.UserGuideCommand;
 import seedu.address.logic.commands.ViewCommand;
 import seedu.address.logic.parser.exceptions.ParseException;
 
+
```
Comment (notle1706) [inline src/main/java/seedu/address/logic/parser/JeeqTrackerParser.java]: Perhaps this line could be deleted.

**Review by notle1706 — APPROVED** [review]
LGTM

**Comment by notle1706** [comment]
Good job!
